index.js

```diff
@@ -22,83 +22,407 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    );
+  }
+}
+
+function checkRegister(register) {
+  if (register !== undefined &&
+      (typeof register !== 'number' || register < 0 || register > 0xff)) {
+    throw new Error('Invalid I2C register ' + register
+      + '. Valid registers are 0 through 0xff.'
+    );
+  }
+}
+
+function checkLength(length) {
+  if (typeof length !== 'number' || length < 0 || length > 32) {
+    throw new Error('Invalid I2C length ' + length
+      + '. Valid lengths are 0 through 32.'
+    );
+  }
+}
+
+function checkCallback(cb) {
+  if (typeof cb !== 'function') {
+    throw new Error('Invalid I2C callback ' + cb);
+  }
+}
+
+function checkBuffer(buffer) {
+  if (!Buffer.isBuffer(buffer) || buffer.length <= 0 || buffer.length > 32) {
+    throw new Error('Invalid I2C buffer ' + buffer
+      + '. Valid lengths are 0 through 32.'
+    );
+  }
+}
+
+function checkByte(byte) {
+  if (typeof byte !== 'number' || byte < 0 || byte > 0xff) {
+    throw new Error('Invalid I2C byte ' + byte
+      + '. Valid values are 0 through 0xff.'
+    );
+  }
+}
+
+function checkWord(word) {
+  if (typeof word !== 'number' || word < 0 || word > 0xffff) {
+    throw new Error('Invalid I2C word ' + word
+      + '. Valid values are 0 through 0xffff.'
+    );
+  }
+}
+
+var devices = '__r$396836_0$__';
+var getDevice = '__r$396836_1$__';
 
 export class I2C extends Peripheral {
   constructor(baudRate, pins) {
     super(pins || [ 'SDA0', 'SCL0' ]);
+
+    Object.defineProperties(this, {
+      [devices]: {
+        writable: true,
+        value: []
+      }
+    });
+
     if (baudRate) {
       if (typeof baudRate != number || baudRate % 1000 != 0) {
         throw new Error('Invalid I2C baud rate. Baud rates must be a multiple of 1000');
       }
-      sh.run('gpio load i2c ' + baudRate);
+      execSync('gpio load i2c ' + baudRate); // Is this still necessary?
     } else {
-      sh.run('gpio load i2c'); // Is this still necessary?
+      execSync('gpio load i2c'); // Is this still necessary?
     }
   }
 
   destroy() {
+    this[devices].forEach(function (device) {
+      device.closeSync();
+    });
+
+    this[devices] = [];
+
     super.destroy();
   }
 
+  [getDevice](address) {
+    var device = this[devices][address];
+
+    if (device === undefined) {
+      device = i2c.openSync(getBoardRevision() === VERSION_1_MODEL_B_REV_1 ? 0 : 1);
+      this[devices][address] = device;
+    }
+
+    return device;
+  }
+
   // function cb(err, buffer), returns undefined, register is optional
+  // Required by J5
   i2cRead(address, register, length, cb) {
-    throw new Error('Not implemented');
+    var buffer;
+    var callback;
+
+    checkAlive(this.alive);
+
+    if (arguments.length === 3) {
+      cb = length;
+      length = register;
+      register = undefined;
+    }
+
+    checkAddress(address);
+    checkRegister(register);
+    checkLength(length);
+    checkCallback(cb);
+
+    buffer = new Buffer(length);
+    callback = function (err) {
```

Comment:
Coding style nits: the only time Raspi.js uses function expressions is when they're inlined, otherwise it should be a function declaration. Also, variables are declared inline, not at the top.